### PR TITLE
typeck: limit number of candidates shown for a single error

### PR DESCRIPTION
**Pull request by birkenfeld**
No idea if 10/11 is a good limit. Are there any other such limits in rustc currently?

Fixes: #25356

### CODE REVIEW

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -236,8 +236,10 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
                          mut sources: Vec<CandidateSource>) {
         sources.sort();
         sources.dedup();
+        // Dynamic limit to avoid hiding just one candidate, which is silly.
+        let limit = if sources.len() == 11 { 11 } else { 10 };
```
Comment (Manishearth) [inline src/librustc_typeck/check/method/suggest.rs]: good idea

Comment (Manishearth) [inline src/librustc_typeck/check/method/suggest.rs]: good idea

**Comment by Manishearth** [comment]
LGTM. I'd like someone else (@jonathandturner?) to rubber-stamp the number 10 as well, though :smile:

**Comment by jonathandturner** [comment]
How difficult would it be to show an elision?

Something like:

```
src\main.rs:206:9: 206:35 error: multiple applicable methods in scope [E0034]
src\main.rs:206     0.0.max(normal.dot(light_dir)) * res
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
src\main.rs:206:9: 206:35 note: candidate #1 is defined in an impl of the trait `core::iter::Iterator` for the type `std::ascii::EscapeDefault`
src\main.rs:206     0.0.max(normal.dot(light_dir)) * res
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
src\main.rs:206:9: 206:35 note: candidate #2 is defined in an impl of the trait `core::iter::Iterator` for the type `std::collections::hash::table::RawBuckets<'_, _, _>`
src\main.rs:206     0.0.max(normal.dot(light_dir)) * res
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
src\main.rs:206:9: 206:35 note: additional candidates elided.
```

FWIW, I'm fine stopping at 10, but I think it'd be good to let the user know there were others, they just aren't shown.

**Comment by birkenfeld** [comment]
@jonathandturner it should already say "N more candidates not shown" at the end.

I just found another example in https://github.com/rust-lang/rust/blob/90318b8c225cde275cf2059330dc096f77301bce/src/librustc/traits/error_reporting.rs#L209 - it stops at 4 and says "and N others" at the end. I could adapt to this.

**Comment by jonathandturner** [comment]
Great, I missed that.  If it has the message, I'm fine stopping even sooner than 10.  4 sounds good.

**Comment by Manishearth** [comment]
sgtm either way. I'd like to switch to "and N others" though, let's keep things uniform.

@bors delegate+

**Comment by Manishearth** [comment]
@bors r- clean force

(something's wrong with the queue)

**Comment by steveklabnik** [comment]
Is the travis failure here legitimate?

**Comment by birkenfeld** [comment]
Yes; the diagnostics API changed on master. I'll fix it soon.

**Comment by birkenfeld** [comment]
Rebased.

**Pull request merged**